### PR TITLE
8280422: thread_from_jni_environment can never return NULL

### PR DESCRIPTION
**Pull request by dholmes-ora**
Analysing the code and its use (see JBS issue for gory details) we can see that `thread_from_jni_environment` can never actually return NULL, so we change it to not appear to do so and thus keep static analysis tools happy.  We also always validate the incoming JNIEnv for debug builds and scrap the unused `VerifyJNIEnvThread` flag.

Testing: tiers 1-5

Thanks,
David

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8280422](https://bugs.openjdk.java.net/browse/JDK-8280422): thread_from_jni_environment can never return NULL

### Reviewers
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**)
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7193/head:pull/7193` \
`$ git checkout pull/7193`

Update a local copy of the PR: \
`$ git checkout pull/7193` \
`$ git pull https://git.openjdk.java.net/jdk pull/7193/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7193`

View PR using the GUI difftool: \
`$ git pr show -t 7193`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7193.diff">https://git.openjdk.java.net/jdk/pull/7193.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/runtime/thread.hpp [hunk 1]
```diff
@@ -1312,16 +1312,21 @@ class JavaThread: public Thread {
   // Returns the jni environment for this thread
   JNIEnv* jni_environment()                      { return &_jni_environment; }
 
+  // Returns the current thread as indicated by the given JNIEnv.
```
Comment (DamonFool) [inline src/hotspot/share/runtime/thread.hpp:1315]: Copyright year: 2021 --> 2022 ?

Comment (dholmes-ora) [reply src/hotspot/share/runtime/thread.hpp:1315]: Fixed. Thanks.

**Review by shipilev — APPROVED** [review]
Looks sensible to me.

**Comment by dholmes-ora** [comment]
Thanks for the review @shipilev !

File: src/hotspot/share/runtime/thread.hpp [hunk 2]
```diff
@@ -1312,16 +1312,21 @@ class JavaThread: public Thread {
...
   static JavaThread* thread_from_jni_environment(JNIEnv* env) {
-    JavaThread *thread_from_jni_env = (JavaThread*)((intptr_t)env - in_bytes(jni_environment_offset()));
-    // Only return NULL if thread is off the thread list; starting to
-    // exit should not return NULL.
-    if (thread_from_jni_env->is_terminated()) {
-      thread_from_jni_env->block_if_vm_exited();
-      return NULL;
-    } else {
-      return thread_from_jni_env;
+    JavaThread* current = (JavaThread*)((intptr_t)env - in_bytes(jni_environment_offset()));
```
Comment (kimbarrett) [inline src/hotspot/share/runtime/thread.hpp:1319]: [pre-existing] If I were writing this I would have used `char*` rather than `intptr_t`.

**Review by kimbarrett — APPROVED** [review]
Looks good.

**Comment by parttimenerd** [comment]
I would suggest to add a method for checking the termination status in cases where it might happen:

```c++
  // Returns whether current thread as indicated by the given JNIEnv
  // is terminated.
  // We don't assert it is Thread::current here as that is done at the
  // external JNI entry points where the JNIEnv is passed into the VM.
  static bool is_thread_from_jni_environment_terminated(JNIEnv* env) {
    JavaThread* current = (JavaThread*)((intptr_t)env - in_bytes(jni_environment_offset()));
    // We can't get here in a thread that has completed its execution and so
    // "is_terminated", but a thread is also considered terminated if the VM
    // has exited, so we have to check this and block in case this is a daemon
    // thread returning to the VM (the JNI DirectBuffer entry points rely on
    // this).
    return current->is_terminated();
  }
```

**Comment by dholmes-ora** [comment]
@parttimenerd 👍 
> This seems to break AsyncGetCallTrace in some rare cases, as "thread_from_jni_environment can never actually return NULL" does not always hold.

Can you please provide further information. As a correctly used JNIEnv can never produce a NULL thread as the JNIEnv is a sub-object of the JavaThread instance.

**Comment by parttimenerd** [comment]
AsyncGetCallTrace is implemented in [forte.cpp](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/prims/forte.cpp):

```cpp
void AsyncGetCallTrace(ASGCT_CallTrace *trace, jint depth, void* ucontext) {
  JavaThread* thread;

  if (trace->env_id == NULL ||
    (thread = JavaThread::thread_from_jni_environment(trace->env_id)) == NULL ||
    thread->is_exiting()) {

    // bad env_id, thread has exited or thread is exiting
    trace->num_frames = ticks_thread_exit; // -8
    return;
  }
```

I'm going to create a separate issue with a PR for the discussion.

**Comment by dholmes-ora** [comment]
That condition should just be:
```
if (trace->env_id == NULL || JavaThread::thread_from_jni_environment(trace->env_id)->is_exiting()) {
```

**Comment by parttimenerd** [comment]
That does not address the issue, as `JavaThread::thread_from_jni_environment(trace->env_id)` never returns in case `is_terminating()` returns true.

**Comment by dholmes-ora** [comment]
Correct but this PR did not introduce that behaviour. I only changed the name of the local variable and added the comment.

File: src/hotspot/share/runtime/thread.hpp [hunk 3]
```diff
@@ -1312,16 +1312,21 @@ class JavaThread: public Thread {
...
+    // We can't get here in a thread that has completed its execution and so
+    // "is_terminated", but a thread is also considered terminated if the VM
+    // has exited, so we have to check this and block in case this is a daemon
+    // thread returning to the VM (the JNI DirectBuffer entry points rely on
+    // this).
+    if (current->is_terminated()) {
+      current->block_if_vm_exited();
+      ShouldNotReachHere();
```
Comment (parttimenerd) [inline src/hotspot/share/runtime/thread.hpp:1327]: This breaks AsyncGetCallTrace

**Comment by parttimenerd** [comment]
It seems to me that this commit introduces the `ShouldNotReachHere()` when before `NULL` would be returned.

File: src/hotspot/share/runtime/thread.hpp (hunk 3 above)
Comment (dholmes-ora) [reply src/hotspot/share/runtime/thread.hpp:1327]: If we reach that code then the VM aborts. It should be impossible to reach that statement and thus impossible to actually return NULL in the old code. Please elaborate under what conditions you think this code can be reached.

Comment (parttimenerd) [reply src/hotspot/share/runtime/thread.hpp:1327]: AsyncProfiler on a debug build of OpenJDK profiling the dacapo tomcat benchmark.

**Comment by parttimenerd** [comment]
I've created a PR with a fix at https://github.com/openjdk/jdk/pull/7559

**Pull request closed without merging**